File: core/src/com/alta_v2/game/actor/overlay/FadeCalculator.java
package com.alta_v2.game.actor.overlay;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * Provides the descriptor for fade in transition.
 */
@RequiredArgsConstructor
class FadeCalculator {

    private final float duration;
    private final OverlayAnimationType overlayAnimationType;

    private float currentDuration;

    @Getter
    private float alpha;

    @Getter
    private boolean completed;

    /**
     * Acts the step in the calculation.
     *
     * @param delta - the delta between rendering.
     */
    void act(float delta) {
        switch (this.overlayAnimationType) {
            case FADE_IN:
                this.calculateFadeIn(delta);
                break;
            case FADE_OUT:
                this.calculateFadeOut(delta);
                break;
        }
    }

    private void calculateFadeIn(float delta) {
        if (this.completed) {
            return;
        }

        this.currentDuration += delta;

        if (this.currentDuration >= duration) {
            this.alpha = 0.0f;
            this.completed = true;
            return;
        }

        // calculates the percentage of current time
        float currentPercentage = this.duration / this.currentDuration;

        // calculates the alpha value by given percentage.
        this.alpha = 1.0f - 1.0f / currentPercentage;
        if (this.alpha <= 0.0f) {
            this.alpha = 0.0f;
            this.completed = true;
        }
    }

    private void calculateFadeOut(float delta) {
        if (this.completed) {
            return;
        }

        this.currentDuration += delta;

        if (this.currentDuration >= duration) {
            this.alpha = 1.0f;
            this.completed = true;
            return;
        }

        // calculates the percentage of current time
        float currentPercentage = this.duration / this.currentDuration;

        // calculates the alpha value by given percentage.
        this.alpha = 1.0f / currentPercentage;

        if (this.alpha >= 1.0f) {
            this.alpha = 1.0f;
            this.completed = true;
        }
    }
}

File: core/src/com/alta_v2/game/overlay/FadeCalculator.kt
package com.alta_v2.game.overlay

/**
 * Provides the descriptor for fade in transition.
 */
internal class FadeCalculator {

    private var currentDuration = 0f

    var duration: Float = 0.5f
    var animationType: OverlayAnimationType = OverlayAnimationType.FADE_IN
    private var alpha = 0f
    private var completed = false

    /**
     * Acts the step in the calculation.
     *
     * @param delta - the delta between rendering.
     */
    fun act(delta: Float) {
        when (animationType) {
            OverlayAnimationType.FADE_IN -> calculateFadeIn(delta)
            OverlayAnimationType.FADE_OUT -> calculateFadeOut(delta)
        }
    }

    fun reset() {
        alpha = 0f
        currentDuration = 0f
        completed = false
    }

    fun isCompleted() = completed

    fun getAlpha() = alpha

    private fun calculateFadeIn(delta: Float) {
        if (completed) {
            return
        }
        currentDuration += delta
        if (currentDuration >= duration) {
            alpha = 0.0f
            completed = true
            return
        }

        // calculates the percentage of current time
        val currentPercentage = duration / currentDuration

        // calculates the alpha value by given percentage.
        alpha = 1.0f - 1.0f / currentPercentage
        if (alpha <= 0.0f) {
            alpha = 0.0f
            completed = true
        }
    }

    private fun calculateFadeOut(delta: Float) {
        if (completed) {
            return
        }
        currentDuration += delta
        if (currentDuration >= duration) {
            alpha = 1.0f
            completed = true
            return
        }

        // calculates the percentage of current time
        val currentPercentage = duration / currentDuration

        // calculates the alpha value by given percentage.
        alpha = 1.0f / currentPercentage
        if (alpha >= 1.0f) {
            alpha = 1.0f
            completed = true
        }
    }
}